TS-4468: http.server_session_sharing.match check SNI

shinrich opened the pull request:
Started with patch proposed by Jered Floyd on the bug.  Tested the basic case of SNI name match/mismatch and reuse seem to work as expected.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/765/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/661/ for details.

jpeach commented on proxy/http/HttpSessionManager.cc:
```diff
@@ -75,17 +75,32 @@ ServerSessionPool::match(HttpServerSession *ss, sockaddr const *addr, INK_MD5 co
+  const char *req_host    = sm->t_state.hdr_info.server_request.host_get(&len);
+  const char *session_sni = netvc->options.sni_servername;
+
+  return ((sm->t_state.scheme != URL_WKSIDX_HTTPS) || !session_sni || !strncmp(session_sni, req_host, len));
```
What is the scheme check doing?

jpeach replied:
``strcmp(...) == 0`` is easier to read IMHO

jpeach replied:
Do you need to make a case-insensitive comparison here?

shinrich replied:
If the scheme is not https, there is no need to do the SNI check.  
Agreed, might as well do a case insensitive compare here.

jpeach commented on proxy/http/HttpSessionManager.cc:
```diff
@@ -98,7 +113,10 @@ ServerSessionPool::acquireSession(sockaddr const *addr, INK_MD5 const &hostname_
     if (TS_SERVER_SESSION_SHARING_MATCH_IP != match_style) {
-      while (loc && loc->hostname_hash != hostname_hash) {
+      while (loc) {
+        if (loc->hostname_hash == hostname_hash && match_sni(sm, loc->get_netvc())) {
```
I don't think this is the right approach. ``match_sni`` should do exactly what it's name claims and we should check in the caller whether we are looking for a TLS session. I think that having ``match_sni`` make assumptions about the match is spooky action at a distance that makes this harder to understand than is necessary.

shinrich replied:
The original patch did not have match_sni.   But the code was cut-n-pasted several places.  I'd rather not pollute the caller with more details.  Could change the function name to be less confusion.  Like match_sni_if_needed.

shinrich commented:
Pushed new version to address the case insensitive name comparison.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/783/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/679/ for details.

SolidWallOfCode commented:
Looks good but needs better comments. The validity of the checking depends on subtle features and interactions that aren't mentioned (most importantly, that the SNI name is tweaked to the remapped FQDN when an origin session is originally created so the request FQDN and SNI are equivalent).

atsci commented:
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/687/ for details.

atsci commented:
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/791/ for details.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/792/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/688/ for details.

shinrich commented:
Changed the name of match_sni to validat_sni.  Added comments in valid_sni to clarify what is being compared and what constraint we are enforcing on connection reuse.
